### PR TITLE
Mec172x gpio driver

### PR DESCRIPTION
**Pull request by scottwcpg**
Add a Version 2 GPIO driver for MEC172x to make future use of the MEC172x ECIA interrupt aggregator driver when that driver is ready. Currently this driver is the same as the MEC152x GPIO driver. Ran GPIO driver basic API and 1pin tests with a temporary hacks (not in this PR).

### CODE REVIEW

**Comment by cfriedt** [comment]
@scottwcpg - can you please squash your relevant commit into one?

See https://docs.zephyrproject.org/latest/contribute/index.html#other-commit-expectations

Based on this and your previous PR's it appears that you are branching from branches when making PR's rather than branching from main. Please ensure you branch from main.

See https://docs.zephyrproject.org/latest/contribute/index.html#contribution-workflow

**Review by cfriedt — CHANGES_REQUESTED** [review]
Commit guidelines & workflow must be followed.

Typically we see at most 2 commits per PR: 1 commit for code, and 1 commit for tests.

**Comment by ravinprv** [comment]
@cfriedt could you point  how to create multiple PR's for  master branch , if we have one open PR1 already from development branch to Master. Assuming each PR(s) has not yet merged to Master and Each PR has it's own branch in same hierarchies ie. 
PR1, PR2...PRn are 1st child branch from Master. 
this will help to follow the process in our next  PR's. 

thanks
ravin

**Comment by cfriedt** [comment]
> @cfriedt could you point  how to create multiple PR's for  master branch 

@ravinprv 
```
git checkout main
git fetch && git pull
git checkout -b pr1
# make some commits
# squash changes with git rebase -i commit_id
# create PR1 on GitHub
git checkout main
git fetch && git pull
git checkout -b pr2
# make some commits
# squash changes with git rebase -i commit_id
# create PR2 on GitHub
```

**Comment by ravinprv** [comment]
@cfriedt thx 
Please reconfirm before we apply your suggested commands; we could create PR2.
PR1 could be still in Review it's NOT MERGED into Main before we create PR2. 

Squashing and rbase is good idea, but still i am thinking main branch will be same as before PR1 as there is no changes
merged either PR1 or PR2.. doesn't GitHub merge PR1 and PR2 commits to One PR?

As i am definitly new comer to Git type SCM, perforce what i use where the process directly works on main repo there is no local commit concepts for local offline repo...

**Comment by scottwcpg** [comment]
@cfriedt Squashed and rebased from main branch (I hope).

**Comment by cfriedt** [comment]
> @cfriedt thx
> Please reconfirm before we apply your suggested commands; we could create PR2.

I'm not suggesting you use the commands I wrote as-written, but it's definitely necessary to follow the commit guidelines linked above.

> PR1 could be still in Review it's NOT MERGED into Main before we create PR2.

Definitely do not rely on a previous & unrelated PR being merged. We like to consider that each PR is a self-contained unit that has no external dependencies. Also, for the most part, we prefer to have all commits in a single PR "squashed" into one commit.

> Squashing and rbase is good idea, but still i am thinking main branch will be same as before PR1 as there is no changes
> merged either PR1 or PR2.. doesn't GitHub merge PR1 and PR2 commits to One PR?

Currently, we require the author to squash commits, and PRs are not merged somehow on GitHub.

> As i am definitly new comer to Git type SCM, perforce what i use where the process directly works on main repo there is no local commit concepts for local offline repo...

I understand - I was there are one point too. What I would suggest doing, is playing around with `git rebase -i` in an "experiment" branch. So just take the current [feature branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) that you are on, do a `git checkout -b experimental`, and try it out! Your previous feature branch will still have all commits preserved.

There are lots of resources out there for learning git, but this one is probably quite useful as well.
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by scottwcpg** [comment]
@cfriedt OK. Self-contained PR's means lots of duplication.  These PR's are all dependent.  These changes are similar to adding a new chip because the HAL is not wanted.

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+	.parin_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_140_176), 1),
+	.parout_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_140_176), 2),
+	.port_num = MCHP_GPIO_140_176,
+#if DT_IRQ_HAS_CELL(DT_NODELABEL(gpio_140_176), irq)
+	.girq_id = 8,
+	.flags = GPIO_INT_ENABLE,
+#else
+	.flags = 0,
```
Comment (jackrosenthal) [inline drivers/gpio/gpio_mchp_xec_v2.c:558]: it seems this code is heavily repeated ... would suggest making it more maintainable by reducing the repeated code ... you can do this using a macro if there really are no other options

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h [hunk 2]
```diff
@@ -1156,4 +1158,80 @@ struct ecia_regs {
...
+{
+	if ((girq < MCHP_FIRST_GIRQ_NOS) || (girq > MCHP_LAST_GIRQ_NOS)) {
+		return 0u;
+	}
+
+	struct ecia_regs *ecia = (struct ecia_regs *)(ECIA_BASE_ADDR);
+
+	return ecia->GIRQ[girq - 8u].RESULT;
+}
+
```
Comment (jackrosenthal) [inline soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:1235]: remove double blank line

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h [hunk 3]
```diff
@@ -10,6 +10,8 @@
 #include <stdint.h>
 #include <stddef.h>
 
+#define ECIA_BASE_ADDR			0x4000e000u
```
Comment (jackrosenthal) [inline soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:13]: this should come from device tree

**Review by jackrosenthal — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 4]
```diff
@@ -0,0 +1,690 @@
...
+
+#define XEC_ECIA_REGS ((struct ecia_regs *) (DT_REG_ADDR(DT_NODELABEL(ecia))))
+
+static const uint32_t valid_ctrl_masks[NUM_MCHP_GPIO_PORTS] = {
+	(MCHP_GPIO_PORT_A_BITMAP),
+	(MCHP_GPIO_PORT_B_BITMAP),
+	(MCHP_GPIO_PORT_C_BITMAP),
+	(MCHP_GPIO_PORT_D_BITMAP),
+	(MCHP_GPIO_PORT_E_BITMAP),
+	(MCHP_GPIO_PORT_F_BITMAP)
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 5]
```diff
@@ -0,0 +1,690 @@
...
+	}
+
+	/* Disable interrupt in the EC aggregator */
+	mchp_soc_ecia_girq_src_dis(config->girq_id, pin);
+
+	/* pin configuration matches requested detection mode? */
+	pcr1 = sys_read32(pcr1_addr);
+
+	if ((pcr1 & MCHP_GPIO_CTRL_IDET_MASK) == pcr1_req) {
+		goto gp_config_girq;
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 6]
```diff
@@ -0,0 +1,690 @@
...
+	.parin_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_000_036), 1),
+	.parout_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_000_036), 2),
+	.port_num = MCHP_GPIO_000_036,
+#if DT_IRQ_HAS_CELL(DT_NODELABEL(gpio_000_036), irq)
+	.girq_id = 11,
+	.flags = GPIO_INT_ENABLE,
+#else
+	.flags = 0,
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 7]
```diff
@@ -0,0 +1,690 @@
...
+	.parin_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_040_076), 1),
+	.parout_addr =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_040_076), 2),
+	.port_num = MCHP_GPIO_040_076,
+#if DT_IRQ_HAS_CELL(DT_NODELABEL(gpio_040_076), irq)
+	.girq_id = 10,
+	.flags = GPIO_INT_ENABLE,
+#else
+	.flags = 0,
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c (hunk 1 above)
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done. Used DT FOREACH and a macro.

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h (hunk 2 above)
Comment (scottwcpg) [reply soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:1236]: Done

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h (hunk 3 above)
Comment (scottwcpg) [reply soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+#include <soc.h>
+
+#include "gpio_utils.h"
+
+#define XEC_GPIO_EDGE_DLY_COUNT		8
+/* read only register for dummy writes */
+#define XEC_GPIO_DLY_ADDR		\
+	(DT_REG_ADDR(DT_NODELABEL(pcr)) + MCHP_PCR_CLK_REQ0_OFS)
+
+#define XEC_ECIA_REGS ((struct ecia_regs *) (DT_REG_ADDR(DT_NODELABEL(ecia))))
```
Comment (jackrosenthal) [inline drivers/gpio/gpio_mchp_xec_v2.c:21]: no space after cast

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 9]
```diff
@@ -0,0 +1,439 @@
...
+	.port_toggle_bits = gpio_xec_port_toggle_bits,
+	.pin_interrupt_configure = gpio_xec_pin_interrupt_configure,
+	.manage_callback = gpio_xec_manage_callback,
+};
+
+#define XEC_GPIO_PORT_FLAGS(n)						\
+	(DT_IRQ_HAS_CELL(DT_DRV_INST(n), irq)) ? GPIO_INT_ENABLE : 0
+
+#define XEC_GPIO_PORT(n)						\
+	static int gpio_xec_port_init_##n(const struct device *dev);	\
```
Comment (jackrosenthal) [inline drivers/gpio/gpio_mchp_xec_v2.c:397]: if you put the definition above the usage, you won't need the forward declaration

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 10]
```diff
@@ -0,0 +1,439 @@
...
+	.port_set_masked_raw = gpio_xec_port_set_masked_raw,
+	.port_set_bits_raw = gpio_xec_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_xec_port_clear_bits_raw,
+	.port_toggle_bits = gpio_xec_port_toggle_bits,
+	.pin_interrupt_configure = gpio_xec_pin_interrupt_configure,
+	.manage_callback = gpio_xec_manage_callback,
+};
+
+#define XEC_GPIO_PORT_FLAGS(n)						\
+	(DT_IRQ_HAS_CELL(DT_DRV_INST(n), irq)) ? GPIO_INT_ENABLE : 0
```
Comment (jackrosenthal) [inline drivers/gpio/gpio_mchp_xec_v2.c:394]: macros with complex values must be enclosed by parenthesis

File: drivers/gpio/gpio_mchp_xec_v2.c (hunk 8 above)
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c (hunk 9 above)
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c (hunk 10 above)
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

File: drivers/gpio/gpio_mchp_xec_v2.c [hunk 11]
```diff
@@ -0,0 +1,690 @@
...
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(gpio_000_036), okay)
+static int gpio_xec_port000_036_init(const struct device *dev);
+
+static const struct gpio_xec_config gpio_xec_port000_036_config = {
+	.common = {
+		.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_NODE(
+			DT_NODELABEL(gpio_000_036)),
+	},
+	.pcr1_base =
+		(uintptr_t) DT_REG_ADDR_BY_IDX(DT_NODELABEL(gpio_000_036), 0),
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply drivers/gpio/gpio_mchp_xec_v2.c]: Done

**Comment by scottwcpg** [comment]
@jackrosenthal @cfriedt I found a bug I created and fixed it.  Also, the suggestion of using ARM __DMB() from 37138 instead of the delay loop worked.

**Review by jackrosenthal — APPROVED** [review]

**Comment by albertofloyd** [comment]
@scottwcpg Could you elaborate why need to duplicate MEC15xx code instead extending existing driver?  
I understand the that this v2 driver uses new introduces ECIA driver concept, is that the only reason? 
i.e. are any other underlying HW blocks differences or other architectural differences that will prevent these 2 functionality to coexist in a single driver?

**Comment by scottwcpg** [comment]
@albertofloyd Multiple reasons
1. MEC172x has an aggregated interrupt driver. I thought interrupt driver could be used to simplify the GPIO driver but the GPIO driver is using a linked list for pin interrupts and processing multiple pins together.  The two driver might not work well together with more hard work.
2. If I am forced to move to a prototype PINCTRL then I do not want to touch MEC152x.
3. I don't want to fight reviewers over changes affecting MEC152x.  If I could I would make ALL MEC172x drivers different.

File: drivers/gpio/Kconfig.xec [hunk 12]
```diff
@@ -5,6 +5,12 @@
 
 config GPIO_XEC
 	bool "XEC Microchip GPIO driver"
-	depends on SOC_FAMILY_MEC
+	depends on SOC_SERIES_MEC1501X
 	help
 	  Enable the Microchip XEC gpio driver.
+
+config GPIO_XEC_V2
```
Comment (albertofloyd) [inline drivers/gpio/Kconfig.xec:12]: @scottwcpg to be considered in upcoming drivers using GPIO_XEC_ECIA instead V2 to distinguish both GPIO approaches

File: soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h [hunk 13]
```diff
@@ -1156,4 +1159,79 @@ struct ecia_regs {
...
+
+static inline uint32_t mchp_soc_ecia_girq_result(uint8_t girq)
+{
+	if ((girq < MCHP_FIRST_GIRQ_NOS) || (girq > MCHP_LAST_GIRQ_NOS)) {
+		return 0u;
+	}
+
+	struct ecia_regs *ecia = (struct ecia_regs *)(ECIA_BASE_ADDR);
+
+	return ecia->GIRQ[girq - 8u].RESULT;
```
Comment (albertofloyd) [inline soc/arm/microchip_mec/mec172x/reg/mec172x_ecia.h:1234]: @scottwcpg Perhaps a macro will make this clearer

**Review by albertofloyd — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**